elide-graphql/src/main/java/com/yahoo/elide/graphql/parser/VariableResolver.java

```diff
@@ -0,0 +1,138 @@
+import graphql.language.StringValue;
+import graphql.language.Type;
+import graphql.language.Value;
+import graphql.language.VariableDefinition;
+import graphql.language.VariableReference;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import javax.ws.rs.BadRequestException;
+
+/**
+ * Class that contains variables provided in graphql request and can resolve variables based on
+ * {@link graphql.language.OperationDefinition} scope.
+ */
+class VariableResolver {
+    private final Map<String, Object> requestVariables;  // variables defined in request is global
+    private Map<String, Object> scopeVariables;  // variables defined in each operation is operation-scoped
+
+    VariableResolver(Map<String, Object> variables) {
+        this.requestVariables = new HashMap<>(variables);
+    }
+
+    /**
+     * Start a new variable scope for operation.
+     *
+     * @param operation operation definition
+     */
+    void newScope(OperationDefinition operation) {
+        this.scopeVariables = new HashMap<>(requestVariables);
+        addVariables(operation.getVariableDefinitions());
+    }
+
+    /**
+     * Resolve {@link VariableDefinition}s and store results in the variable map.
+     *
+     * @param definitions definitions to resolve
+     */
+    private void addVariables(List<VariableDefinition> definitions) {
+        definitions.stream()
+                .filter(variable -> variable.getDefaultValue() != null
+                        && variable.getDefaultValue() instanceof VariableReference)
+                .forEach(this::addVariable);
+    }
+
+    /**
+     * Resolve {@link VariableDefinition} and store result in the variable map.
+     * We don't need to worry about resolving graphql {@link graphql.language.TypeName} here because Elide-core
+     * knows the correct type of each field/argument.
+     *
+     * @param definition definition to resolve
+     */
+    private void addVariable(VariableDefinition definition) {
+        Type variableType = definition.getType();
+        String variableName = definition.getName();
+        Value defaultValue = definition.getDefaultValue();
+
+        if (defaultValue == null) {
+            if (variableType instanceof NonNullType && scopeVariables.get(variableName) == null) {
+                // value of non-null variable must be resolvable
+                throw new BadRequestException("Undefined non-null variable " + variableName);
+            } else {
+                // this would put 'null' for this variable if it is not stored in the map
+                scopeVariables.put(variableName, scopeVariables.get(variableName));
+            }
+        } else {
+            if (!scopeVariables.containsKey(variableName)) {
+                // create a new variable with default value
+                scopeVariables.put(variableName, resolveValue(defaultValue));
+            } else {
```

Comment:
I don't understand what is going in this list scenario.

Reply:
In `GraphQLEndpointTest.testValidFetchWithVariables()`, we defined `bookId` variable with value `1` and that is referenced in the `book.ids` argument. The `book.ids` argument is a list, but `1` is just an object.